app/code/Magento/InventoryCatalog/Plugin/CatalogInventory/Model/ResourceModel/UpdateSourceItemAtLegacyCatalogInventoryQtyCounter.php

```diff
@@ -0,0 +1,147 @@
+     */
+    private $sourceItemsSave;
+
+    /**
+     * @var DefaultSourceProviderInterface
+     */
+    private $defaultSourceProvider;
+
+    /**
+     * @var SearchCriteriaBuilder
+     */
+    private $searchCriteriaBuilder;
+
+    /**
+     * @var ResourceConnection
+     */
+    private $resourceConnection;
+
+    /**
+     * @param ProductRepositoryInterface $productRepository
+     * @param SourceItemRepositoryInterface $sourceItemRepository
+     * @param SourceItemsSaveInterface $sourceItemsSave
+     * @param DefaultSourceProviderInterface $defaultSourceProvider
+     * @param SearchCriteriaBuilder $searchCriteriaBuilder
+     * @param ResourceConnection $resourceConnection
+     */
+    public function __construct(
+        ProductRepositoryInterface $productRepository,
+        SourceItemRepositoryInterface $sourceItemRepository,
+        SourceItemsSaveInterface $sourceItemsSave,
+        DefaultSourceProviderInterface $defaultSourceProvider,
+        SearchCriteriaBuilder $searchCriteriaBuilder,
+        ResourceConnection $resourceConnection
+    ) {
+        $this->productRepository = $productRepository;
+        $this->sourceItemRepository = $sourceItemRepository;
+        $this->sourceItemsSave = $sourceItemsSave;
+        $this->defaultSourceProvider = $defaultSourceProvider;
+        $this->searchCriteriaBuilder = $searchCriteriaBuilder;
+        $this->resourceConnection = $resourceConnection;
+    }
+
+    /**
+     * @param $subject
+     * @param callable $proceed
+     * @param int[] $items
+     * @param int $websiteId
+     * @param string $operator +/-
+     *
+     * @return void
+     * @throws \Exception
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function aroundCorrectItemsQty($subject, callable $proceed, array $items, $websiteId, $operator)
+    {
+        $proceed($items, $websiteId, $operator);
+
+        if ($websiteId !== 0) {
+            return;
+        }
+
+        $connection = $this->resourceConnection->getConnection();
+        $connection->beginTransaction();
+
+        try {
+            $productsData = $this->getProductData($items);
+
+            $searchCriteria =
+                $this->searchCriteriaBuilder->addFilter(SourceItemInterface::SKU, array_keys($productsData), 'in')
+                                            ->addFilter(
+                                                SourceItemInterface::SOURCE_ID,
+                                                $this->defaultSourceProvider->getId()
+                                            )
+                                            ->create();
+
+            $sourceItems = $this->sourceItemRepository->getList($searchCriteria)->getItems();
+            $sourceItems = array_map(
+                function (SourceItemInterface $item) use ($productsData, $operator) {
+                    $item->setQuantity($item->getQuantity() + (int)($operator . $productsData[$item->getSku()]));
```

Comment:
why do we make casting to (int) but not to (float) ? 
looks like we have unneeded rounding